docs/gaslessWithERC4337.md

```diff
@@ -0,0 +1,131 @@
+- **calculating and transferring**: after executing the transaction, the entrypoint will calculate how much gas fee the transaction costs, and take the responsive cost from the paymaster balance(deposit in the entrypoint's fund pool, remember?) to the block producer.
+
+### Workflow
+
+When a user signs a gasless transaction, the transaction is not only with 0 gas price but also supposed to call the Entrypoint contracts with a call data that specifies what they really want to do. That call data uses a data structure named `UserOperation` and it contains info like the real gas price that the sponsor is paying, the paymaster contract address, the real target contract that the user is calling, etc.
+
+```solidity
+struct UserOperation {
+    address callContract;
+    bytes   callData;
+    uint256 callGasLimit;
+    uint256 verificationGasLimit;
+    uint256 preVerificationGas;
+    uint256 maxFeePerGas;
+    uint256 maxPriorityFeePerGas;
+    bytes   paymasterAndData;
+}
+```
+
+Entrypoint will take this `UserOperation` call data to call the paymaster contract to validate if such a sponsor transaction is legal. If it is, entrypoint will charge the gas fee and execute the transaction. If it is not, the transaction shall revert. 
+More details about how the Godwoken node assumes a transaction is a gasless transaction:
+
+1. `tx.gasPrice` must be `0`.
+2. `tx.data` must be a valid `UserOperation`. 
+3. `tx.gas` must equal to **UserOperation.callGasLimit + UserOperation.verificationGasLimit * 3**.
+4. `tx.to` must equal to `EntryPoint`'s address.
+5. `UserOperation.paymasterAndData` must contain a valid paymaster address.
+
+With users able to send free-charge transactions and block_producer able to receive funds from sponsor dApps, we can finally bring a gasless transaction experience to mainstream users.
+
+### Difference between Gasless feature and ERC-4337
+
+For those who are familiar with ERC-4337 proposal, you might already find entrypoint and paymaster contracts quite ring a bell. In fact, the gasless feature is intended to use a simplified version of ERC-4337 specs. 
+The main difference is that ERC-4337 is also an account abstraction focus, which is to make mainstream users get rid of private keys and mnemonic stuff. You might often hear this as "smart-contract wallet". 
+
+In a high-level explanation, 
+> ERC-4337 = seedless feature + gasless feature
+
+While in Godwoken, we only do the last half of it. We believe it is a better choice for Godwoken to abort the seedless idea in the current phase, due to many reasons:
+- ERC-4337 is still in the protocol design phase, it is not widely accepted by the community. It might take a while before it makes a real influence on crypto. However, we want the users and developers to enjoy the gasless feature today.
```

Comment:
```suggestion
- ERC-4337 is still in the protocol design phase, it is not widely accepted by the community. It might take a while before it makes a real influence in the crypto industry. However, we want the users and developers to enjoy the gasless feature today.
```